### ISSUE TITLE
Seal traits inside `impl_` submodule

### ISSUE THREAD

**Issue by davidhewitt** [PyO3/pyo3#3908]
Off the back of https://github.com/PyO3/pyo3/pull/3897#discussion_r1504721475

Should we review the traits inside the `impl_` submodule, check which ones can be sealed, and do that? It's a reasonable safety mechanism to prevent users from trying to implement those things.

**Comment by davidhewitt** [comment]
I labelled this "Good First Issue", as this is an internal refactoring which would be great to have.

A "sealed" trait has a private supertrait, thus making it possible to implement outside the crate. All of the `PyAnyMethods` and similar traits are sealed, for example. Take a look at `src/sealed.rs` as the trait we use as the supertrait for most sealing.

**Comment by davidhewitt** [comment]
Brilliant, thanks 🙏

**Comment by Cheukting** [comment]
@davidhewitt quick question, is there any particular reason something should be or should not be moved? I want to make sure I fully understand this.

**Comment by davidhewitt** [comment]
I don't think we should be moving any code, this issue is about upgrading `trait Foo` to `trait Foo: Sealed` where it makes sense to do so for the traits inside the `impl_` submodule.

**Comment by Cheukting** [comment]
I see, so those traits that could be used by the user need to be `Sealed` right? Thanks for clarifying it @davidhewitt

**Comment by davidhewitt** [comment]
Yes, in principle many things inside `impl_` are accessible by users but we never intend them to use them directly (just via the macros). The worry is there might be traits in there which we would prefer users can't implement at all, hence the sealing.

**Comment by Cheukting** [comment]
Thank you for your patience, I was occupied by EuroPython preparation in the last few weeks. I plan to pick this one up again this weekend or next week.

**Comment by davidhewitt** [comment]
I think the next step along here would be finding traits which make use of the `private_impl!` helper and changing them all to be `Sealed` instead (and remove the old helper).

**Comment by Cheukting** [comment]
@davidhewitt @LilyFoote is this issue closed? It seems so to me

**Comment by davidhewitt** [comment]
I think there are still lots of traits in `impl_` (search for `pub trait`) which do not have a `Sealed` bound. So it looks like there is more which can be done.

**Comment by Cheukting** [comment]
Cool, I will go treasure hunt when I got the time!

**Comment by luxedo** [comment]
I have found these:

```
$ rg -P -U 'pub trait (?!.*Sealed|.*\n.*Sealed)' src/impl_
src/impl_/callback.rs
10:pub trait PyCallbackOutput: Copy {
28:pub trait IntoPyCallbackOutput<'py, Target> {
122:pub trait WrappingCastTo<T> {

src/impl_/pyclass/probes.rs
17:pub trait Probe {

src/impl_/pyclass/doc.rs
9:pub trait PyClassNewTextSignature {

src/impl_/pyfunction.rs
42:pub trait WrapPyFunctionArg<'py, T> {

src/impl_/trampoline.rs
34:pub trait MethodDef<T> {

src/impl_/pyclass_init.rs
60:pub trait PyClassInit<'py, const IS_PYCLASS: bool, const IS_INITIALIZER_TUPLE: bool> {

src/impl_/introspection.rs
7:pub trait PyReturnType {

src/impl_/pyclass.rs
158:pub trait PyClassImpl: Sized + 'static {
308:        pub trait $trait_name<T>: Sized {
899:pub trait PyClassWithFreeList: PyClass {
999:pub trait PyClassInventory: inventory::Collect {
1006:pub trait PyMethods<T> {
1112:pub trait PyClassBaseType: Sized {
1306:pub trait PyO3GetField<'py>: IntoPyObject<'py> + Clone {}
1444:pub trait ExtractPyClassWithClone {}

src/impl_/wrap.rs
8:pub trait SomeWrap<T> {

src/impl_/pymethods.rs
484:pub trait IterBaseKind {
512:pub trait IterOptionKind {
542:pub trait IterResultOptionKind {
565:pub trait AsyncIterBaseKind {
593:pub trait AsyncIterOptionKind {
623:pub trait AsyncIterResultOptionKind {

src/impl_/extract_argument.rs
155:pub trait FunctionArgumentHolder: Sized {
778:pub trait VarargsHandler<'py> {
856:pub trait VarkeywordsHandler<'py> {
```

- [x] callback.rs https://github.com/PyO3/pyo3/pull/5821
- [x] pyclass/probes.rs https://github.com/PyO3/pyo3/pull/5806
- [x] pyclass/doc.rs - Won't seal
- [x] pyfunction.rs https://github.com/PyO3/pyo3/pull/5806
- [x] trampoline.rs - Won't seal. I couldn't target the `struct Def` without setting the seal module private
- [x] pyclass_init.rs https://github.com/PyO3/pyo3/pull/5844
- [x] introspection.rs https://github.com/PyO3/pyo3/pull/5791
- [x] pyclass.rs https://github.com/PyO3/pyo3/pull/5845
- [x] wrap.rs - Won't seal. Too generic
```rust
impl<T> SomeWrap<T> for T {
impl<T> SomeWrap<T> for Option<T> {
```
- [x] pymethods.rs - Won't seal. Too generic
```rust
impl<Value> IterBaseKind for &Value {}
impl<Value> IterOptionKind for Option<Value> {}
impl<Value, Error> IterResultOptionKind for Result<Option<Value>, Error> {}
impl<Value> AsyncIterBaseKind for &Value {}
impl<Value> AsyncIterOptionKind for Option<Value> {}
impl<Value, Error> AsyncIterResultOptionKind for Result<Option<Value>, Error> {}
```
- [x] extract_argument.rs https://github.com/PyO3/pyo3/pull/5804
